In [21]:
import pandas as pd 
import numpy as np

In [22]:
movies = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\movies.csv" , index_col = 0)

In [23]:
movies

id     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
5     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
4736  48363          0           3      321952   
4743  48370      27000          19     3151130   
4748  48375          0           7           0   
4749  48376          0           3           0   
4768  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
5                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
4736                            The Last Waltz           7.9          64   
4743                                    Clerks           7.4         755   
4748                                   Rampage           6.0         131   
4749                                   Slacker           6.4          77   
4768                               El Mariachi           6.6         238   

      director_id  year month       day  
0            4762  2009   Dec  Thursday  
1            4763  2007   May  Saturday  
2            4764  2015   Oct    Monday  
3            4765  2012   Jul    Monday  
5            4767  2007   May   Tuesday  
...           ...   ...   ...       ...  
4736         4809  1978   May    Monday  
4743         5369  1994   Sep   Tuesday  
4748         5148  2009   Aug    Friday  
4749         5535  1990   Jul    Friday  
4768         5097  1992   Sep    Friday  

[1465 rows x 11 columns]

In [24]:
directors = pd.read_csv(r"C:\Users\DELL\Desktop\Scaler Numpy and Pandas\directors.csv" , index_col = 0)

In [25]:
directors

director_name    id gender
0         James Cameron  4762   Male
1        Gore Verbinski  4763   Male
2            Sam Mendes  4764   Male
3     Christopher Nolan  4765   Male
4        Andrew Stanton  4766   Male
...                 ...   ...    ...
2344      Shane Carruth  7106   Male
2345   Neill Dela Llana  7107    NaN
2346        Scott Smith  7108    NaN
2347        Daniel Hsia  7109   Male
2348   Brian Herzlinger  7110   Male

[2349 rows x 3 columns]

In [26]:
movies.head(1)

id     budget  popularity     revenue   title  vote_average  vote_count  \
0  43597  237000000         150  2787965087  Avatar           7.2       11800   

   director_id  year month       day  
0         4762  2009   Dec  Thursday

In [27]:
directors.head(1)

director_name    id gender
0  James Cameron  4762   Male

In [28]:
data = movies.merge(directors , left_on = "director_id" , right_on = "id" , how = "left")

In [29]:
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_y gender  
0            4762  2009   Dec  Thursday      James Cameron  4762   Male  
1            4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2            4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3            4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4            4767  2007   May   Tuesday          Sam Raimi  4767   Male  
...           ...   ...   ...       ...                ...   ...    ...  
1460         4809  1978   May    Monday    Martin Scorsese  4809   Male  
1461         5369  1994   Sep   Tuesday        Kevin Smith  5369   Male  
1462         5148  2009   Aug    Friday           Uwe Boll  5148   Male  
1463         5535  1990   Jul    Friday  Richard Linklater  5535   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez  5097    NaN  

[1465 rows x 14 columns]

In [30]:
data.drop(["director_id" , "id_y"] , axis = 1 , inplace = True)

In [31]:
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

In [32]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   int64  
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   int64  
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   year           1465 non-null   int64  
 8   month          1465 non-null   object 
 9   day            1465 non-null   object 
 10  director_name  1465 non-null   object 
 11  gender         1341 non-null   object 
dtypes: float64(1), int64(6), object(5)
memory usage: 148.8+ KB


In [33]:
data.describe(include = "all")

id_x        budget   popularity       revenue   title  \
count    1465.000000  1.465000e+03  1465.000000  1.465000e+03    1465   
unique           NaN           NaN          NaN           NaN    1465   
top              NaN           NaN          NaN           NaN  Avatar   
freq             NaN           NaN          NaN           NaN       1   
mean    45225.191126  4.802295e+07    30.855973  1.432539e+08     NaN   
std      1189.096396  4.935541e+07    34.845214  2.064918e+08     NaN   
min     43597.000000  0.000000e+00     0.000000  0.000000e+00     NaN   
25%     44236.000000  1.400000e+07    11.000000  1.738013e+07     NaN   
50%     45022.000000  3.300000e+07    23.000000  7.578164e+07     NaN   
75%     45990.000000  6.600000e+07    41.000000  1.792469e+08     NaN   
max     48395.000000  3.800000e+08   724.000000  2.787965e+09     NaN   

        vote_average    vote_count         year month     day  \
count    1465.000000   1465.000000  1465.000000  1465    1465   
unique           NaN           NaN          NaN    12       7   
top              NaN           NaN          NaN   Dec  Friday   
freq             NaN           NaN          NaN   193     654   
mean        6.368191   1146.396587  2002.615017   NaN     NaN   
std         0.818033   1578.077438     8.680141   NaN     NaN   
min         3.000000      1.000000  1976.000000   NaN     NaN   
25%         5.900000    216.000000  1998.000000   NaN     NaN   
50%         6.400000    571.000000  2004.000000   NaN     NaN   
75%         6.900000   1387.000000  2009.000000   NaN     NaN   
max         8.300000  13752.000000  2016.000000   NaN     NaN   

           director_name gender  
count               1465   1341  
unique               199      2  
top     Steven Spielberg   Male  
freq                  26   1309  
mean                 NaN    NaN  
std                  NaN    NaN  
min                  NaN    NaN  
25%                  NaN    NaN  
50%                  NaN    NaN  
75%                  NaN    NaN  
max                  NaN    NaN

In [34]:
data.head(10)

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   
5  43606  250000000         155   873260194   
6  43607  270000000          57   391081192   
7  43608  200000000         107   586090727   
8  43609  200000000         145  1065659812   
9  43610  255000000          49    89289910   

                                        title  vote_average  vote_count  year  \
0                                      Avatar           7.2       11800  2009   
1    Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                     Spectre           6.3        4466  2015   
3                       The Dark Knight Rises           7.6        9106  2012   
4                                Spider-Man 3           5.9        3576  2007   
5          Batman v Superman: Dawn of Justice           5.7        7004  2016   
6                            Superman Returns           5.4        1400  2006   
7                           Quantum of Solace           6.1        2965  2008   
8  Pirates of the Caribbean: Dead Man's Chest           7.0        5246  2006   
9                             The Lone Ranger           5.9        2311  2013   

  month        day      director_name gender  
0   Dec   Thursday      James Cameron   Male  
1   May   Saturday     Gore Verbinski   Male  
2   Oct     Monday         Sam Mendes   Male  
3   Jul     Monday  Christopher Nolan   Male  
4   May    Tuesday          Sam Raimi   Male  
5   Mar  Wednesday        Zack Snyder   Male  
6   Jun  Wednesday       Bryan Singer   Male  
7   Oct   Thursday       Marc Forster   Male  
8   Jun    Tuesday     Gore Verbinski   Male  
9   Jul  Wednesday     Gore Verbinski   Male

# In the above data if we see the revenue, it is the amount in dollars. So for better undestanding I do want to change it to millions of dollars. we can do it in the way as shown below.

In [35]:
data["revenue"] = (data["revenue"]/1000000).round(2)

In [36]:
data

id_x     budget  popularity  revenue  \
0     43597  237000000         150  2787.97   
1     43598  300000000         139   961.00   
2     43599  245000000         107   880.67   
3     43600  250000000         112  1084.94   
4     43602  258000000         115   890.87   
...     ...        ...         ...      ...   
1460  48363          0           3     0.32   
1461  48370      27000          19     3.15   
1462  48375          0           7     0.00   
1463  48376          0           3     0.00   
1464  48395     220000          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

# So after doing the above operation the revenue is changed to millions. example - In the first row the revenue is 2787.97 million dollars.

# Also I want to change the budget in the same way to millions of dollars.

In [37]:
data.head(2)

id_x     budget  popularity  revenue  \
0  43597  237000000         150  2787.97   
1  43598  300000000         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [38]:
data["budget"] = (data["budget"]/1000000).round(2)

In [39]:
data

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

# Question 1 - Find all the high rated movies, which are rated above 7.

In [40]:
data.head(5)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

# Method 1

In [41]:
data[data["vote_average"] > 7] 

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
3     43600  250.00         112  1084.94   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
19    43622  200.00         100  1845.03   
...     ...     ...         ...      ...   
1456  48321    0.01          20     7.00   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name gender  
0     2009   Dec   Thursday      James Cameron   Male  
3     2012   Jul     Monday  Christopher Nolan   Male  
14    2014   Dec  Wednesday      Peter Jackson   Male  
16    2013   Dec  Wednesday      Peter Jackson   Male  
19    1997   Nov    Tuesday      James Cameron   Male  
...    ...   ...        ...                ...    ...  
1456  1977   Mar   Saturday        David Lynch   Male  
1457  1998   Oct     Friday      Peter Chelsom   Male  
1458  1998   Jul     Friday   Darren Aronofsky   Male  
1460  1978   May     Monday    Martin Scorsese   Male  
1461  1994   Sep    Tuesday        Kevin Smith   Male  

[301 rows x 12 columns]

# Method -2 (Using loc)

In [42]:
data.loc[data["vote_average"] > 7]

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
3     43600  250.00         112  1084.94   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
19    43622  200.00         100  1845.03   
...     ...     ...         ...      ...   
1456  48321    0.01          20     7.00   
1457  48323    0.00           5     0.00   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   

                                          title  vote_average  vote_count  \
0                                        Avatar           7.2       11800   
3                         The Dark Knight Rises           7.6        9106   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
19                                      Titanic           7.5        7562   
...                                         ...           ...         ...   
1456                                 Eraserhead           7.5         485   
1457                                 The Mighty           7.1          51   
1458                                         Pi           7.1         586   
1460                             The Last Waltz           7.9          64   
1461                                     Clerks           7.4         755   

      year month        day      director_name gender  
0     2009   Dec   Thursday      James Cameron   Male  
3     2012   Jul     Monday  Christopher Nolan   Male  
14    2014   Dec  Wednesday      Peter Jackson   Male  
16    2013   Dec  Wednesday      Peter Jackson   Male  
19    1997   Nov    Tuesday      James Cameron   Male  
...    ...   ...        ...                ...    ...  
1456  1977   Mar   Saturday        David Lynch   Male  
1457  1998   Oct     Friday      Peter Chelsom   Male  
1458  1998   Jul     Friday   Darren Aronofsky   Male  
1460  1978   May     Monday    Martin Scorsese   Male  
1461  1994   Sep    Tuesday        Kevin Smith   Male  

[301 rows x 12 columns]

# Copy the original data 

In [43]:
df = data.copy()

In [44]:
df

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  
1461  1994   Sep   Tuesday        Kevin Smith   Male  
1462  2009   Aug    Friday           Uwe Boll   Male  
1463  1990   Jul    Friday  Richard Linklater   Male  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  

[1465 rows x 12 columns]

# Question 2 - Find all the movies that are rated above 7 and extract only the title of the movie and the director name.

In [45]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [46]:
df.loc[df["vote_average"] > 7 , ["title" , "director_name"]]

title      director_name
0                                        Avatar      James Cameron
3                         The Dark Knight Rises  Christopher Nolan
14    The Hobbit: The Battle of the Five Armies      Peter Jackson
16          The Hobbit: The Desolation of Smaug      Peter Jackson
19                                      Titanic      James Cameron
...                                         ...                ...
1456                                 Eraserhead        David Lynch
1457                                 The Mighty      Peter Chelsom
1458                                         Pi   Darren Aronofsky
1460                             The Last Waltz    Martin Scorsese
1461                                     Clerks        Kevin Smith

[301 rows x 2 columns]

# Question 3 - Find out all the movies rated above 7 and released in or after the year 2015.

In [47]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [48]:
df.loc[(df["vote_average"] > 7) & (df["year"] >= 2015)]

id_x  budget  popularity  revenue                    title  \
30    43641   190.0         102  1506.25                Furious 7   
78    43724   150.0         434   378.86       Mad Max: Fury Road   
106   43773   135.0         100   532.95             The Revenant   
162   43867   108.0         167   630.16              The Martian   
312   44128    75.0          48   108.15  The Man from U.N.C.L.E.   
394   44281    44.0          68   155.76        The Hateful Eight   
625   44770    35.0          53   194.56               The Intern   
635   44784    40.0          48   165.48          Bridge of Spies   
808   45194    30.0          65    91.71                 Southpaw   
833   45293    28.0          61   201.63   Straight Outta Compton   
839   45301    28.0          57   133.35            The Big Short   
1344  47181     5.0          22    24.80                     Race   

      vote_average  vote_count  year month        day  \
30             7.3        4176  2015   Apr  Wednesday   
78             7.2        9427  2015   May  Wednesday   
106            7.3        6396  2015   Dec     Friday   
162            7.6        7268  2015   Sep  Wednesday   
312            7.1        2265  2015   Aug   Thursday   
394            7.6        4274  2015   Dec     Friday   
625            7.1        1881  2015   Sep   Thursday   
635            7.2        2583  2015   Oct   Thursday   
808            7.3        2067  2015   Jun     Monday   
833            7.7        1355  2015   Aug   Thursday   
839            7.3        2607  2015   Dec     Friday   
1344           7.1         478  2016   Feb     Friday   

                    director_name  gender  
30                      James Wan    Male  
78                  George Miller    Male  
106   Alejandro González Iñárritu    Male  
162                  Ridley Scott    Male  
312                   Guy Ritchie    Male  
394             Quentin Tarantino    Male  
625                  Nancy Meyers  Female  
635              Steven Spielberg    Male  
808                 Antoine Fuqua    Male  
833                  F. Gary Gray    Male  
839                    Adam McKay    Male  
1344              Stephen Hopkins    Male

# Question 4 - Find all the movies which are alphabetically before the movie title avengers.

In [49]:
df.loc[df["title"] < "Avengers"]

id_x  budget  popularity  revenue                           title  \
0     43597   237.0         150  2787.97                          Avatar   
23    43629   200.0          78  1025.49             Alice in Wonderland   
40    43656   200.0          45   769.65                            2012   
41    43657   200.0          39   325.23               A Christmas Carol   
69    43709   155.0          39   167.30                       Alexander   
...     ...     ...         ...      ...                             ...   
1390  47491     3.0           7    20.97              A Room with a View   
1395  47575     3.0           3     0.00                        Amnesiac   
1405  47686     2.0          23    20.91                   Amores perros   
1432  47970     0.0           3     0.00              All the Real Girls   
1440  48155     0.0           0     0.00  Alleluia! The Devil's Carnival   

      vote_average  vote_count  year month        day  \
0              7.2       11800  2009   Dec   Thursday   
23             6.4        4645  2010   Mar  Wednesday   
40             5.6        4903  2009   Oct   Saturday   
41             6.6        1095  2009   Nov  Wednesday   
69             5.6         927  2004   Nov     Sunday   
...            ...         ...   ...   ...        ...   
1390           6.9         156  1985   Dec     Friday   
1395           4.1          52  2015   Aug     Friday   
1405           7.6         521  2000   Jun     Friday   
1432           5.9          30  2003   Aug     Friday   
1440           6.0           2  2016   Mar    Tuesday   

                    director_name gender  
0                   James Cameron   Male  
23                     Tim Burton   Male  
40                Roland Emmerich   Male  
41                Robert Zemeckis   Male  
69                   Oliver Stone   Male  
...                           ...    ...  
1390                  James Ivory   Male  
1395               Michael Polish   Male  
1405  Alejandro González Iñárritu   Male  
1432           David Gordon Green   Male  
1440          Darren Lynn Bousman   Male  

[125 rows x 12 columns]

# Question 5 - Find all the movies where there is a word "the".

In [50]:
df.loc[df["title"].str.contains("The")]

id_x  budget  popularity  revenue  \
3     43600  250.00         112  1084.94   
9     43610  255.00          49    89.29   
11    43612  225.00          53   419.65   
14    43616  250.00         120   956.02   
16    43619  250.00          94   958.40   
...     ...     ...         ...      ...   
1440  48155    0.00           0     0.00   
1443  48192    0.35          35    29.40   
1449  48244    0.25           6     0.06   
1457  48323    0.00           5     0.00   
1460  48363    0.00           3     0.32   

                                          title  vote_average  vote_count  \
3                         The Dark Knight Rises           7.6        9106   
9                               The Lone Ranger           5.9        2311   
11     The Chronicles of Narnia: Prince Caspian           6.3        1630   
14    The Hobbit: The Battle of the Five Armies           7.1        4760   
16          The Hobbit: The Desolation of Smaug           7.6        4524   
...                                         ...           ...         ...   
1440             Alleluia! The Devil's Carnival           6.0           2   
1443                              The Evil Dead           7.3         894   
1449                                The Canyons           4.1          75   
1457                                 The Mighty           7.1          51   
1460                             The Last Waltz           7.9          64   

      year month        day        director_name gender  
3     2012   Jul     Monday    Christopher Nolan   Male  
9     2013   Jul  Wednesday       Gore Verbinski   Male  
11    2008   May   Thursday       Andrew Adamson   Male  
14    2014   Dec  Wednesday        Peter Jackson   Male  
16    2013   Dec  Wednesday        Peter Jackson   Male  
...    ...   ...        ...                  ...    ...  
1440  2016   Mar    Tuesday  Darren Lynn Bousman   Male  
1443  1981   Oct   Thursday            Sam Raimi   Male  
1449  2013   Jul     Monday        Paul Schrader    NaN  
1457  1998   Oct     Friday        Peter Chelsom   Male  
1460  1978   May     Monday      Martin Scorsese   Male  

[327 rows x 12 columns]

# Question 6 - Search all the movies where the name of the movie starts with the word "batman"

In [51]:
df.loc[df["title"].str.startswith("Batman")]

id_x  budget  popularity  revenue                               title  \
5    43606   250.0         155   873.26  Batman v Superman: Dawn of Justice   
74   43716   150.0         115   374.22                       Batman Begins   
128  43807   125.0          50   238.21                      Batman & Robin   
184  43896   100.0          48   336.53                      Batman Forever   
257  44025    80.0          59   280.00                      Batman Returns   
704  44956    35.0          44   411.35                              Batman   

     vote_average  vote_count  year month        day      director_name gender  
5             5.7        7004  2016   Mar  Wednesday        Zack Snyder   Male  
74            7.5        7359  2005   Jun     Friday  Christopher Nolan   Male  
128           4.2        1418  1997   Jun     Friday    Joel Schumacher   Male  
184           5.2        1498  1995   May  Wednesday    Joel Schumacher   Male  
257           6.6        1673  1992   Jun     Friday         Tim Burton   Male  
704           7.0        2096  1989   Jun     Friday         Tim Burton   Male

# Question 7 - Find out the top most popular movies.

In [52]:
df.head(5)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                   Spectre           6.3        4466  2015   
3                     The Dark Knight Rises           7.6        9106  2012   
4                              Spider-Man 3           5.9        3576  2007   

  month       day      director_name gender  
0   Dec  Thursday      James Cameron   Male  
1   May  Saturday     Gore Verbinski   Male  
2   Oct    Monday         Sam Mendes   Male  
3   Jul    Monday  Christopher Nolan   Male  
4   May   Tuesday          Sam Raimi   Male

In [53]:
df.sort_values("popularity" , ascending = False).head(10)

id_x  budget  popularity  revenue  \
58   43692   165.0         724   675.12   
78   43724   150.0         434   378.86   
119  43796   140.0         271   655.01   
120  43797   125.0         206   752.10   
45   43662   185.0         187  1004.56   
162  43867   108.0         167   630.16   
59   43693   160.0         167   825.53   
5    43606   250.0         155   873.26   
0    43597   237.0         150  2787.97   
383  44259    63.0         146   100.85   

                                                 title  vote_average  \
58                                        Interstellar           8.1   
78                                  Mad Max: Fury Road           7.2   
119  Pirates of the Caribbean: The Curse of the Bla...           7.5   
120              The Hunger Games: Mockingjay - Part 1           6.6   
45                                     The Dark Knight           8.2   
162                                        The Martian           7.6   
59                                           Inception           8.1   
5                   Batman v Superman: Dawn of Justice           5.7   
0                                               Avatar           7.2   
383                                         Fight Club           8.3   

     vote_count  year month        day      director_name gender  
58        10867  2014   Nov  Wednesday  Christopher Nolan   Male  
78         9427  2015   May  Wednesday      George Miller   Male  
119        6985  2003   Jul  Wednesday     Gore Verbinski   Male  
120        5584  2014   Nov    Tuesday   Francis Lawrence   Male  
45        12002  2008   Jul  Wednesday  Christopher Nolan   Male  
162        7268  2015   Sep  Wednesday       Ridley Scott   Male  
59        13752  2010   Jul  Wednesday  Christopher Nolan   Male  
5          7004  2016   Mar  Wednesday        Zack Snyder   Male  
0         11800  2009   Dec   Thursday      James Cameron   Male  
383        9413  1999   Oct     Friday      David Fincher   Male

# Question 8 - Find all the movies created by Cristopher Nolan.

In [54]:
df.loc[df["director_name"] == "Christopher Nolan" , ["title"]]

title
3     The Dark Knight Rises
45          The Dark Knight
58             Interstellar
59                Inception
74            Batman Begins
565                Insomnia
641            The Prestige
1341                Memento

# Question 9 - Number of movies released by a particular director Christopher Nolan.

In [55]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

# Method - 1

In [56]:
df.loc[df["director_name"] == "Christopher Nolan" , ["title"]].nunique()

title    8
dtype: int64

# Method - 2

In [57]:
df.loc[df["director_name"] == "Christopher Nolan" , ["title"]].count()

title    8
dtype: int64

# Question 10 - Find the number of of movies directed by all the directors.

In [58]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [59]:
df["director_name"].value_counts()

Steven Spielberg      26
Martin Scorsese       19
Clint Eastwood        19
Woody Allen           18
Ridley Scott          16
                      ..
Tim Hill               5
Jonathan Liebesman     5
Roman Polanski         5
Larry Charles          5
Nicole Holofcener      5
Name: director_name, Length: 199, dtype: int64

# ------- Grouping ------

In [60]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [61]:
df.groupby(["director_name"])

# Question 10 - Calculate the maximum budget of the directors.

In [62]:
df.head(1)

id_x  budget  popularity  revenue   title  vote_average  vote_count  year  \
0  43597   237.0         150  2787.97  Avatar           7.2       11800  2009   

  month       day  director_name gender  
0   Dec  Thursday  James Cameron   Male

In [63]:
df.groupby(["director_name"])["budget"].max()

director_name
Adam McKay                     100.0
Adam Shankman                   80.0
Alejandro González Iñárritu    135.0
Alex Proyas                    140.0
Alexander Payne                 30.0
                               ...  
Wes Craven                      40.0
Wolfgang Petersen              175.0
Woody Allen                     30.0
Zack Snyder                    250.0
Zhang Yimou                     94.0
Name: budget, Length: 199, dtype: float64

# Question 11 - Calculate the maximum revenue of the directors.

In [64]:
df.head(1)

id_x  budget  popularity  revenue   title  vote_average  vote_count  year  \
0  43597   237.0         150  2787.97  Avatar           7.2       11800  2009   

  month       day  director_name gender  
0   Dec  Thursday  James Cameron   Male

In [65]:
df.groupby(["director_name"])["revenue"].max()

director_name
Adam McKay                     173.65
Adam Shankman                  212.87
Alejandro González Iñárritu    532.95
Alex Proyas                    347.23
Alexander Payne                177.24
                                ...  
Wes Craven                     173.05
Wolfgang Petersen              497.41
Woody Allen                    151.12
Zack Snyder                    873.26
Zhang Yimou                    177.39
Name: revenue, Length: 199, dtype: float64

# Question 12 - Number of movies created by each director.

In [66]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [67]:
df.groupby(["director_name"])["title"].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

# sorting the above.

In [68]:
df.groupby(["director_name"])["title"].count().sort_values()

director_name
Kevin Reynolds       5
George Clooney       5
Gavin O'Connor       5
Gary Winick          5
Michael Apted        5
                    ..
Spike Lee           16
Woody Allen         18
Clint Eastwood      19
Martin Scorsese     19
Steven Spielberg    26
Name: title, Length: 199, dtype: int64

In [69]:
df.groupby(["director_name"])["title"].count().sort_values(ascending = False)

director_name
Steven Spielberg    26
Clint Eastwood      19
Martin Scorsese     19
Woody Allen         18
Robert Rodriguez    16
                    ..
Paul Weitz           5
John Madden          5
Paul Verhoeven       5
John Whitesell       5
Kevin Reynolds       5
Name: title, Length: 199, dtype: int64

# Question 13 - Find out the most Productive director. (Director who have produced most number of movies per year).

In [70]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [71]:
df.groupby(["director_name"])["year"].min()

director_name
Adam McKay                     2004
Adam Shankman                  2001
Alejandro González Iñárritu    2000
Alex Proyas                    1994
Alexander Payne                1999
                               ... 
Wes Craven                     1984
Wolfgang Petersen              1981
Woody Allen                    1977
Zack Snyder                    2004
Zhang Yimou                    2002
Name: year, Length: 199, dtype: int64

In [72]:
df.groupby(["director_name"])["year"].max()

director_name
Adam McKay                     2015
Adam Shankman                  2012
Alejandro González Iñárritu    2015
Alex Proyas                    2016
Alexander Payne                2013
                               ... 
Wes Craven                     2011
Wolfgang Petersen              2006
Woody Allen                    2013
Zack Snyder                    2016
Zhang Yimou                    2014
Name: year, Length: 199, dtype: int64

In [73]:
df.groupby(["director_name"])["title"].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

In [74]:
df_agg = df.groupby(["director_name"])[["year" , "title"]].aggregate({"title" : "count" , "year" : ["min" , "max"]})

In [75]:
df_agg

title  year      
                            count   min   max
director_name                                
Adam McKay                      6  2004  2015
Adam Shankman                   8  2001  2012
Alejandro González Iñárritu     6  2000  2015
Alex Proyas                     5  1994  2016
Alexander Payne                 5  1999  2013
...                           ...   ...   ...
Wes Craven                     10  1984  2011
Wolfgang Petersen               7  1981  2006
Woody Allen                    18  1977  2013
Zack Snyder                     7  2004  2016
Zhang Yimou                     6  2002  2014

[199 rows x 3 columns]

In [76]:
df_agg.columns

MultiIndex([('title', 'count'),
            ( 'year',   'min'),
            ( 'year',   'max')],
           )

In [77]:
df_agg.columns = ["_".join(col) for col in df_agg.columns]

In [78]:
df_agg

title_count  year_min  year_max
director_name                                               
Adam McKay                             6      2004      2015
Adam Shankman                          8      2001      2012
Alejandro González Iñárritu            6      2000      2015
Alex Proyas                            5      1994      2016
Alexander Payne                        5      1999      2013
...                                  ...       ...       ...
Wes Craven                            10      1984      2011
Wolfgang Petersen                      7      1981      2006
Woody Allen                           18      1977      2013
Zack Snyder                            7      2004      2016
Zhang Yimou                            6      2002      2014

[199 rows x 3 columns]

In [79]:
df_agg.reset_index(inplace = True)

In [80]:
df_agg

director_name  title_count  year_min  year_max
0                     Adam McKay            6      2004      2015
1                  Adam Shankman            8      2001      2012
2    Alejandro González Iñárritu            6      2000      2015
3                    Alex Proyas            5      1994      2016
4                Alexander Payne            5      1999      2013
..                           ...          ...       ...       ...
194                   Wes Craven           10      1984      2011
195            Wolfgang Petersen            7      1981      2006
196                  Woody Allen           18      1977      2013
197                  Zack Snyder            7      2004      2016
198                  Zhang Yimou            6      2002      2014

[199 rows x 4 columns]

In [81]:
df_agg["years_active"] = df_agg["year_max"] - df_agg["year_min"]

In [82]:
df_agg

director_name  title_count  year_min  year_max  \
0                     Adam McKay            6      2004      2015   
1                  Adam Shankman            8      2001      2012   
2    Alejandro González Iñárritu            6      2000      2015   
3                    Alex Proyas            5      1994      2016   
4                Alexander Payne            5      1999      2013   
..                           ...          ...       ...       ...   
194                   Wes Craven           10      1984      2011   
195            Wolfgang Petersen            7      1981      2006   
196                  Woody Allen           18      1977      2013   
197                  Zack Snyder            7      2004      2016   
198                  Zhang Yimou            6      2002      2014   

     years_active  
0              11  
1              11  
2              15  
3              22  
4              14  
..            ...  
194            27  
195            25  
196            36  
197            12  
198            12  

[199 rows x 5 columns]

In [83]:
df_agg["movies_per_year"] = df_agg["title_count"] / df_agg["years_active"]

In [84]:
df_agg

director_name  title_count  year_min  year_max  \
0                     Adam McKay            6      2004      2015   
1                  Adam Shankman            8      2001      2012   
2    Alejandro González Iñárritu            6      2000      2015   
3                    Alex Proyas            5      1994      2016   
4                Alexander Payne            5      1999      2013   
..                           ...          ...       ...       ...   
194                   Wes Craven           10      1984      2011   
195            Wolfgang Petersen            7      1981      2006   
196                  Woody Allen           18      1977      2013   
197                  Zack Snyder            7      2004      2016   
198                  Zhang Yimou            6      2002      2014   

     years_active  movies_per_year  
0              11         0.545455  
1              11         0.727273  
2              15         0.400000  
3              22         0.227273  
4              14         0.357143  
..            ...              ...  
194            27         0.370370  
195            25         0.280000  
196            36         0.500000  
197            12         0.583333  
198            12         0.500000  

[199 rows x 6 columns]

In [85]:
df_agg.sort_values(["movies_per_year"] , ascending = False)

director_name  title_count  year_min  year_max  years_active  \
190       Tyler Perry            9      2006      2013             7   
73    Jason Friedberg            5      2006      2010             4   
169        Shawn Levy           11      2002      2014            12   
158  Robert Rodriguez           16      1992      2014            22   
1       Adam Shankman            8      2001      2012            11   
..                ...          ...       ...       ...           ...   
104   Lawrence Kasdan            5      1985      2012            27   
109        Luc Besson            5      1985      2014            29   
157    Robert Redford            5      1980      2010            30   
170      Sidney Lumet            5      1976      2006            30   
117     Michael Apted            5      1980      2010            30   

     movies_per_year  
190         1.285714  
73          1.250000  
169         0.916667  
158         0.727273  
1           0.727273  
..               ...  
104         0.185185  
109         0.172414  
157         0.166667  
170         0.166667  
117         0.166667  

[199 rows x 6 columns]

In [86]:
df_agg.sort_values(["movies_per_year"] , inplace = True , ascending = False)

In [87]:
df_agg.head(10)

director_name  title_count  year_min  year_max  years_active  \
190       Tyler Perry            9      2006      2013             7   
73    Jason Friedberg            5      2006      2010             4   
169        Shawn Levy           11      2002      2014            12   
158  Robert Rodriguez           16      1992      2014            22   
1       Adam Shankman            8      2001      2012            11   
156    Robert Luketic            6      2001      2010             9   
179  Steven Spielberg           26      1977      2016            39   
8        Andy Fickman            6      2006      2015             9   
112      Marc Forster            8      2001      2013            12   
81    Joel Schumacher           13      1987      2007            20   

     movies_per_year  
190         1.285714  
73          1.250000  
169         0.916667  
158         0.727273  
1           0.727273  
156         0.666667  
179         0.666667  
8           0.666667  
112         0.666667  
81          0.650000

# Question 14 - Find the maximum budget of all the directors.

In [88]:
df.head(2)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   

                                      title  vote_average  vote_count  year  \
0                                    Avatar           7.2       11800  2009   
1  Pirates of the Caribbean: At World's End           6.9        4500  2007   

  month       day   director_name gender  
0   Dec  Thursday   James Cameron   Male  
1   May  Saturday  Gore Verbinski   Male

In [89]:
df_dir_budget = df.groupby(["director_name"])["budget"].max().reset_index()

In [90]:
df_dir_budget

director_name  budget
0                     Adam McKay   100.0
1                  Adam Shankman    80.0
2    Alejandro González Iñárritu   135.0
3                    Alex Proyas   140.0
4                Alexander Payne    30.0
..                           ...     ...
194                   Wes Craven    40.0
195            Wolfgang Petersen   175.0
196                  Woody Allen    30.0
197                  Zack Snyder   250.0
198                  Zhang Yimou    94.0

[199 rows x 2 columns]

# Question 15 -  Find out the directors that created atleast one movie which is greater than or equal to 100 million.

In [91]:
df_dir_budget.loc[df_dir_budget["budget"] >= 100]

director_name  budget
0                     Adam McKay   100.0
2    Alejandro González Iñárritu   135.0
3                    Alex Proyas   140.0
5                 Andrew Adamson   225.0
10                       Ang Lee   137.0
..                           ...     ...
187                  Tom Shadyac   175.0
188                   Tom Tykwer   102.0
189                   Tony Scott   100.0
195            Wolfgang Petersen   175.0
197                  Zack Snyder   250.0

[85 rows x 2 columns]

In [92]:
df.groupby(["director_name"]).filter(lambda x:x["budget"].max() >= 100)

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1450  48267    0.40          33   100.00   
1451  48268    0.20          13     4.51   
1452  48274    0.00           5     2.61   
1458  48335    0.06          27     3.22   
1460  48363    0.00           3     0.32   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1450                                   Mad Max           6.6        1213   
1451                                  Swingers           6.8         253   
1452                                     Three           6.3          31   
1458                                        Pi           7.1         586   
1460                            The Last Waltz           7.9          64   

      year month       day      director_name gender  
0     2009   Dec  Thursday      James Cameron   Male  
1     2007   May  Saturday     Gore Verbinski   Male  
2     2015   Oct    Monday         Sam Mendes   Male  
3     2012   Jul    Monday  Christopher Nolan   Male  
4     2007   May   Tuesday          Sam Raimi   Male  
...    ...   ...       ...                ...    ...  
1450  1979   Apr  Thursday      George Miller   Male  
1451  1996   Oct    Friday         Doug Liman   Male  
1452  2010   Dec  Thursday         Tom Tykwer   Male  
1458  1998   Jul    Friday   Darren Aronofsky   Male  
1460  1978   May    Monday    Martin Scorsese   Male  

[679 rows x 12 columns]

# Question 15 - Find out all the instances where the budget of a particular director for a movie is expensive. The movie is expensive when the budget of the particular movie is more than the average budget of the particular director.

In [93]:
def budget_avg(x):
    x = x - x.mean()
    return x
df.groupby(["director_name"])["budget"].transform(budget_avg)

0       130.300000
1       141.857143
2       150.142857
3       124.375000
4       174.004545
           ...    
1460    -47.478947
1461    -11.976667
1462    -21.700000
1463    -10.890909
1464    -31.168750
Name: budget, Length: 1465, dtype: float64

# Question 16 - Find out the instances of movies where directors are taking a risk. Risk is when revenue is less than the mean value of the buget or budget is more than the mean of revenue for particular director.

In [94]:
data.head(10)

id_x  budget  popularity  revenue  \
0  43597   237.0         150  2787.97   
1  43598   300.0         139   961.00   
2  43599   245.0         107   880.67   
3  43600   250.0         112  1084.94   
4  43602   258.0         115   890.87   
5  43606   250.0         155   873.26   
6  43607   270.0          57   391.08   
7  43608   200.0         107   586.09   
8  43609   200.0         145  1065.66   
9  43610   255.0          49    89.29   

                                        title  vote_average  vote_count  year  \
0                                      Avatar           7.2       11800  2009   
1    Pirates of the Caribbean: At World's End           6.9        4500  2007   
2                                     Spectre           6.3        4466  2015   
3                       The Dark Knight Rises           7.6        9106  2012   
4                                Spider-Man 3           5.9        3576  2007   
5          Batman v Superman: Dawn of Justice           5.7        7004  2016   
6                            Superman Returns           5.4        1400  2006   
7                           Quantum of Solace           6.1        2965  2008   
8  Pirates of the Caribbean: Dead Man's Chest           7.0        5246  2006   
9                             The Lone Ranger           5.9        2311  2013   

  month        day      director_name gender  
0   Dec   Thursday      James Cameron   Male  
1   May   Saturday     Gore Verbinski   Male  
2   Oct     Monday         Sam Mendes   Male  
3   Jul     Monday  Christopher Nolan   Male  
4   May    Tuesday          Sam Raimi   Male  
5   Mar  Wednesday        Zack Snyder   Male  
6   Jun  Wednesday       Bryan Singer   Male  
7   Oct   Thursday       Marc Forster   Male  
8   Jun    Tuesday     Gore Verbinski   Male  
9   Jul  Wednesday     Gore Verbinski   Male

In [95]:
 def func1(x):
        x["risky"] = x["budget"] - x["revenue"].mean()>=0
        return x
df_risky = df.groupby(["director_name"]).apply(func1)
df_risky

id_x  budget  popularity  revenue  \
0     43597  237.00         150  2787.97   
1     43598  300.00         139   961.00   
2     43599  245.00         107   880.67   
3     43600  250.00         112  1084.94   
4     43602  258.00         115   890.87   
...     ...     ...         ...      ...   
1460  48363    0.00           3     0.32   
1461  48370    0.03          19     3.15   
1462  48375    0.00           7     0.00   
1463  48376    0.00           3     0.00   
1464  48395    0.22          14     2.04   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      year month       day      director_name gender  risky  
0     2009   Dec  Thursday      James Cameron   Male  False  
1     2007   May  Saturday     Gore Verbinski   Male  False  
2     2015   Oct    Monday         Sam Mendes   Male  False  
3     2012   Jul    Monday  Christopher Nolan   Male  False  
4     2007   May   Tuesday          Sam Raimi   Male  False  
...    ...   ...       ...                ...    ...    ...  
1460  1978   May    Monday    Martin Scorsese   Male  False  
1461  1994   Sep   Tuesday        Kevin Smith   Male  False  
1462  2009   Aug    Friday           Uwe Boll   Male  False  
1463  1990   Jul    Friday  Richard Linklater   Male  False  
1464  1992   Sep    Friday   Robert Rodriguez    NaN  False  

[1465 rows x 13 columns]

In [96]:
df_risky[df_risky["risky"] == True]

id_x  budget  popularity  revenue  \
7     43608   200.0         107   586.09   
12    43614   380.0         135  1045.71   
15    43618   200.0          37   310.67   
20    43624   209.0          64   303.03   
24    43630   210.0           3   459.36   
...     ...     ...         ...      ...   
1347  47224     5.0           7     3.26   
1349  47229     5.0           3     4.84   
1351  47233     5.0           6     0.00   
1356  47263    15.0          10     0.00   
1383  47453     3.5           4     0.00   

                                            title  vote_average  vote_count  \
7                               Quantum of Solace           6.1        2965   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
15                                     Robin Hood           6.2        1398   
20                                     Battleship           5.5        2114   
24                          X-Men: The Last Stand           6.3        3525   
...                                           ...           ...         ...   
1347                          The Sweet Hereafter           6.8         103   
1349                         90 Minutes in Heaven           5.4          40   
1351                                Light Sleeper           5.7          15   
1356                           Dying of the Light           4.5         118   
1383                  In the Name of the King III           3.3          19   

      year month        day   director_name gender  risky  
7     2008   Oct   Thursday    Marc Forster   Male   True  
12    2011   May   Saturday    Rob Marshall   Male   True  
15    2010   May  Wednesday    Ridley Scott   Male   True  
20    2012   Apr  Wednesday      Peter Berg   Male   True  
24    2006   May  Wednesday    Brett Ratner   Male   True  
...    ...   ...        ...             ...    ...    ...  
1347  1997   May  Wednesday     Atom Egoyan   Male   True  
1349  2015   Sep     Friday  Michael Polish   Male   True  
1351  1992   Aug     Friday   Paul Schrader    NaN   True  
1356  2014   Dec   Thursday   Paul Schrader    NaN   True  
1383  2013   Dec     Friday        Uwe Boll   Male   True  

[131 rows x 13 columns]

In [97]:
df[["revenue" , "budget"]].apply(np.sum , axis = 0)

revenue    209867.04
budget      70353.62
dtype: float64

In [98]:
df[["revenue" , "budget"]].apply(np.sum , axis = 1)

0       3024.97
1       1261.00
2       1125.67
3       1334.94
4       1148.87
         ...   
1460       0.32
1461       3.18
1462       0.00
1463       0.00
1464       2.26
Length: 1465, dtype: float64